src/action/finder/hash.rs

```diff
@@ -0,0 +1,201 @@
+use crate::fs::Entry;
+use crypto::digest::Digest;
+use log::warn;
+use rrg_macro::ack;
+use rrg_proto::file_finder_hash_action_options::OversizedFilePolicy;
+use rrg_proto::Hash as HashEntry;
+use std::cmp::min;
+use std::fs::File;
+use std::io::Read;
+use std::iter::repeat;
+
+/// Hashes data writen to it using SHA-1, SHA-256 and MD5 algorithms.
+struct Hasher {
+    /// Digest with SHA-1 hash.
+    sha1: crypto::sha1::Sha1,
+    /// Digest with SHA-256 hash.
+    sha256: crypto::sha2::Sha256,
+    /// Digest with MD5 hash.
+    md5: crypto::md5::Md5,
+    /// Stores total number of bytes inserted into hasher.
+    total_byte_count: u64,
+}
+
+impl Hasher {
+    pub fn new() -> Hasher {
+        Hasher {
+            sha1: crypto::sha1::Sha1::new(),
+            sha256: crypto::sha2::Sha256::new(),
+            md5: crypto::md5::Md5::new(),
+            total_byte_count: 0,
+        }
+    }
+}
+
+impl std::io::Write for Hasher {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.sha1.input(buf);
+        self.sha256.input(buf);
+        self.md5.input(buf);
+        self.total_byte_count += buf.len() as u64;
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+/// Performs `hash` action on the file in `entry` and returns the result to be reported in case of success.
+pub fn hash(entry: &Entry, config: &HashActionOptions) -> Option<HashEntry> {
+    match config.oversized_file_policy {
+        OversizedFilePolicy::Skip => {
+            if entry.metadata.len() > config.max_size {
+                return None;
+            }
+        }
+        OversizedFilePolicy::HashTruncated => {}
+    };
+
+    let file = ack! {
+        File::open(&entry.path),
+        error: "failed to open file: {}", entry.path.display()
+    }?;
+    let mut file = file.take(config.max_size);
+
+    let mut hasher = Hasher::new();
+    let read_bytes = ack! {
+        std::io::copy(&mut file, &mut hasher),
+        error: "failed to copy data from: {}", entry.path.display()
+    }?;
+
+    let expected_bytes = min(entry.metadata.len(), config.max_size);
+    if read_bytes != expected_bytes {
+        warn!(
+            "failed to read all data from: {}, {} bytes were read, but {} were expected",
+            entry.path.display(),
+            &read_bytes,
+            expected_bytes
+        );
+        return None;
+    }
+
+    Some(HashEntry {
+        sha1: Some(result_vec(&mut hasher.sha1)),
+        sha256: Some(result_vec(&mut hasher.sha256)),
+        md5: Some(result_vec(&mut hasher.md5)),
+        pecoff_sha1: None,
+        pecoff_md5: None,
+        pecoff_sha256: None,
+        signed_data: vec![],
+        num_bytes: Some(hasher.total_byte_count),
+        source_offset: None,
+    })
+}
+
+fn result_vec<T: Digest>(digest: &mut T) -> Vec<u8> {
+    let mut vec = repeat(0).take(digest.output_bytes()).collect::<Vec<u8>>();
```

Comment:
This must be the most convoluted way of writing `vec![0; digest.output_bytes()]` I have seen.

Reply:
Actually idea of doing it that way comes from the crypto library. `vec![0; digest.output_bytes()]` looks way better.